provider/lpseal/task_sdr.go

```diff
@@ -0,0 +1,220 @@
+
+		commd, err = nonffi.GenerateUnsealedCID(sectorParams.RegSealProof, pieceInfos)
+		if err != nil {
+			return false, xerrors.Errorf("computing CommD: %w", err)
+		}
+	} else {
+		commd = zerocomm.ZeroPieceCommitment(abi.PaddedPieceSize(ssize).Unpadded())
+	}
+
+	sref := storiface.SectorRef{
+		ID: abi.SectorID{
+			Miner:  abi.ActorID(sectorParams.SpID),
+			Number: abi.SectorNumber(sectorParams.SectorNumber),
+		},
+		ProofType: sectorParams.RegSealProof,
+	}
+
+	// get ticket
+	maddr, err := address.NewIDAddress(uint64(sectorParams.SpID))
+	if err != nil {
+		return false, xerrors.Errorf("getting miner address: %w", err)
+	}
+
+	// FAIL: api may be down
+	// FAIL-RESP: rely on harmony retry
+	ticket, ticketEpoch, err := s.getTicket(ctx, maddr)
+	if err != nil {
+		return false, xerrors.Errorf("getting ticket: %w", err)
+	}
+
+	// do the SDR!!
+
+	// FAIL: storage may not have enough space
+	// FAIL-RESP: rely on harmony retry
+
+	// LATEFAIL: compute error in sdr
+	// LATEFAIL-RESP: Check in Trees task should catch this; Will retry computing
+	//                Trees; After one retry, it should return the sector to the
+	// 			      SDR stage; max number of retries should be configurable
+
+	err = s.sc.GenerateSDR(ctx, sref, ticket, commd)
+	if err != nil {
+		return false, xerrors.Errorf("generating sdr: %w", err)
+	}
+
+	// store success!
+	n, err := s.db.Exec(ctx, `UPDATE sectors_sdr_pipeline
+		SET after_sdr = true, ticket_epoch = $3, ticket_value = $4
+		WHERE sp_id = $1 AND sector_number = $2`,
+		sectorParams.SpID, sectorParams.SectorNumber, ticketEpoch, []byte(ticket))
+	if err != nil {
+		return false, xerrors.Errorf("store sdr success: updating pipeline: %w", err)
+	}
+	if n != 1 {
+		return false, xerrors.Errorf("store sdr success: updated %d rows", n)
+	}
+
+	return true, nil
+}
+
+func (s *SDRTask) getTicket(ctx context.Context, maddr address.Address) (abi.SealRandomness, abi.ChainEpoch, error) {
+	ts, err := s.api.ChainHead(ctx)
+	if err != nil {
+		return nil, 0, xerrors.Errorf("getting chain head: %w", err)
+	}
+
+	ticketEpoch := ts.Height() - policy.SealRandomnessLookback
+	buf := new(bytes.Buffer)
+	if err := maddr.MarshalCBOR(buf); err != nil {
+		return nil, 0, xerrors.Errorf("marshaling miner address: %w", err)
+	}
+
+	rand, err := s.api.StateGetRandomnessFromTickets(ctx, crypto.DomainSeparationTag_SealRandomness, ticketEpoch, buf.Bytes(), ts.Key())
+	if err != nil {
+		return nil, 0, xerrors.Errorf("getting randomness from tickets: %w", err)
+	}
+
+	return abi.SealRandomness(rand), ticketEpoch, nil
+}
+
+func (s *SDRTask) CanAccept(ids []harmonytask.TaskID, engine *harmonytask.TaskEngine) (*harmonytask.TaskID, error) {
+	// todo check storage (reserve too?)
+
+	id := ids[0]
+	return &id, nil
+}
+
+func (s *SDRTask) TypeDetails() harmonytask.TaskTypeDetails {
+	res := harmonytask.TaskTypeDetails{
+		Max:  s.max,
+		Name: "SDR",
+		Cost: resources.Resources{ // todo offset for prefetch?
+			Cpu: 4, // todo multicore sdr
```

Comment:
It's... If `FIL_PROOFS_USE_MULTICORE_SDR=1` is not set, it's 1, if it is set, that /really/ depends but usually means '1 CCX', except on some never ZEN cores where it can mean 'half of a CCX' or '1/4 of a CCX'

(multicore SDR uses 1-3 threads to prefetch data from RAM into shared L3 cache and one thread to actually run the compute, and it's not easy to tell in Go what number of cores will be used because we don't have access to hwloc and rust code which makes those decisions)

Lotus-miner just makes a guess here, and then it makes this number configurable